File: java/core/src/main/java/com/amazon/controller/PersonController.java
package com.amazon.controller;

import com.amazon.dao.PersonDao;
import com.amazon.mapper.GenericRequest;
import com.amazon.mapper.GenericResponse;
import com.amazon.model.Person;
import com.google.gson.*;
import lombok.extern.slf4j.Slf4j;

import java.time.OffsetDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Map;
import java.util.UUID;

@Slf4j
public class PersonController {

    private static final String GET = "GET";

    private static final String POST = "POST";

    private static final String PERSON_ROUTE = "/person";

    private static final String ID = "id";

    private static final Gson gson = new GsonBuilder()
            .setPrettyPrinting()
            .registerTypeAdapter(
                    OffsetDateTime.class,
                    (JsonDeserializer<OffsetDateTime>) (json, type, context) -> OffsetDateTime.parse(json.getAsString())
            )
            .registerTypeAdapter(
                    OffsetDateTime.class,
                    (JsonSerializer<OffsetDateTime>) (object, type, context) -> new JsonPrimitive(object.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME))
            )
            .create();

    private final PersonDao personDao;

    public PersonController(PersonDao personDao) {
        this.personDao = personDao;
    }

    /**
     * Contains the API routes/handlers - usually based on the request's HTTP method and the URI/path.
     *
     * @param request {@link GenericRequest}
     * @param response {@link GenericResponse}
     */
    public void handleRequest(GenericRequest request, GenericResponse response) {
        // GET /person?id={id}
        if (GET.equals(request.getHttpMethod()) && PERSON_ROUTE.equals(request.getPath())) {
            String id = request.getQueryParameters().get(ID).get(0);  // todo: check if it's okay to get the 0th element.
            UUID uuid = UUID.fromString(id);

            this.personDao.getById(uuid)
                    .ifPresentOrElse(
                            person -> {
                                log.debug("{}", person);

                                response.setStatus(200);

                                String personJson = gson.toJson(person);
                                log.debug(personJson);

                                response.setBody(personJson);
                            },
                            () -> response.setStatus(404)
                    );
        }
        // POST /person
        else if (POST.equals(request.getHttpMethod()) && PERSON_ROUTE.equals(request.getPath())) {
            JsonObject body = request.getBody();

            Person person = gson.fromJson(body, Person.class);
            this.personDao.save(person);

            response.setStatus(201);
            response.setBody(
                    gson.toJson(
                            Map.of(
                                    ID, person.getId().toString()
                            )
                    )
            );
        }
        // If the HTTP method and path cannot be found, then return a 404.
        else {
            log.debug("No endpoint matches {} {}", request.getHttpMethod(), request.getPath());
            response.setStatus(404);
        }

        beforeResponse(request, response);
    }

    /**
     * This method contains logic that should run AFTER the request has been handled, but BEFORE the response is sent back
     * to the client.
     *
     * @param request {@link GenericRequest}
     * @param response {@link GenericResponse}
     */
    private static void beforeResponse(GenericRequest request, GenericResponse response) {
        response.addHeader("Content-Type", "application/json");
    }

}

File: python/mcacore/person_controller.py
from .model import GenericRequest, GenericResponse, Person
from .person_dao import PersonDao


def handle_request(request: GenericRequest, person_dao: PersonDao) -> GenericResponse:
    if request.http_method == 'GET' and request.path == '/person':
        return _get_person(request, person_dao)
    elif request.http_method == 'POST' and request.path == '/person':
        return _create_person(request, person_dao)
    else:
        return build_error_response(request.http_method, 404, 'Not implemented')


def build_error_response(method: str, status: int, msg: str = None) -> GenericResponse:
    return GenericResponse(
        http_method=method,
        status=status,
        body={
            'msg': msg
        }
    )


def _get_person(request: GenericRequest, person_dao: PersonDao) -> GenericResponse:
    try:
        id: str = request.query_parameters.get('id')
        # Get the record from the DAO
        person: Person = person_dao.get_person(id)
        print(f'person is: {person}')
        if not person:
            return build_error_response(request.http_method, 404)

        return GenericResponse(
            http_method=request.http_method,
            status=200,
            body=person.to_json()
        )
    except Exception as exc:
        return build_error_response(request.http_method, 404, f"Not found - {exc}")


def _create_person(request: GenericRequest, person_dao: PersonDao) -> GenericResponse:
    try:
        # Create the person
        person = Person.init_from(json_object=request.body)
        persisted_person = person_dao.create_person(person)

        return GenericResponse(
            http_method=request.http_method,
            status=201,
            body={
                'id': persisted_person.id
            }
        )
    except Exception as exc:
        return build_error_response(request.http_method, 500, f"Internal server error - {exc}")

